### PR TITLE
Add Everest storage (port of seba_sqlite logic)

### LINKED ISSUE #8811: Refactor communication and storage of optimization results
The current storage for everest optimization results should be ported away from the seba-sqlite to something as close to the ERT storage as possible. The results generated by Everest can be seen as:
(1) forward model runs w/ inputs (Everest controls / ERT parameters) and outputs (Everest objectives / ERT GEN_DATA responses)
(2) optimization results computed by ROPT, where the forward model responses are used as inputs. Some of these are not explicitly existing in ERT storage

Course of action:
1. Move storage logic out of seba_sqlite: https://github.com/equinor/ert/pull/9161
2. https://github.com/equinor/ert/issues/9751
3. (blocked by [#9751](https://github.com/equinor/ert/issues/9751)) https://github.com/equinor/ert/issues/9674
4. (blocked by [#9751](https://github.com/equinor/ert/issues/9751))  https://github.com/equinor/ert/issues/9752

Considerations/notes:
* All things computed by ROPT cannot be retrieved from ERT storage, this includes constraint violations, and basically any dataframe starting with `batch_`, meaning various aggregations of results per-realization. Recomputation of these should be very fast if we resolve [#9764](https://github.com/equinor/ert/issues/9674)

**Issue**
Optimization results are currently stored in an SQLite database, which is used by Everest via a snapshot object to report and store results. This approach  is not directly supported anymore by the [`ropt`](https://github.com/tno-ropt/ropt) which is used by Everest to run the optimization. Everest uses the [`seba-sqlite`](https://github.com/tno/seba-sqlite) package which provides the original code from Seba to support the SQLite database. 

This code should be replaced in Everest for the following reasons:

- Reliability issues with the SQLite database, in particular on NFS files systems.
- Performance is poor.
- The code has not kept up with new features that have added in recent years, impeding proper monitoring and storage of results. For instance, when auto-normalization is used, the console output does not handle scaled results properly due to missing information.
- The  `seba-sqlite` code is deprecated, TNO will not continue to support it.

Instead of using the database, Everest should receive optimization results directly from `ropt` via the provided callback mechanism, and take appropriate actions to report and store results.

**Definition of done**
When the SQLite code has been removed from Everest and all current reporting and storage functionality is available.

~~Blocked by https://github.com/equinor/everest/issues/118~~

### PR DESCRIPTION
**Pull request by yngve-sk**
Release notes:
* Stores constraint violation values
* Replaces seba sqlite storage

(Storage PR back in, for reviews, previous PR (accidentally merged): https://github.com/equinor/ert/pull/9161)

**Issue**
Resolves #8811

Base idea/documentation: 
====================
Store datasets by `[batch, realization, perturbation] x [controls, objectives, constraints, objective_gradient, constraint_gradient]`:

Exhaustive list of data stored **PER BATCH** :
===================================
* `batch.json` - contains info about the batch, batch_id and whether it is an improvement (aka merit flag, but the concepts are now unified for dakota and non-dakota runs)
* `batch_constraints` constraint values (and violations) for constraints, batch-wide
* `batch_objectives` objective values, batch-wide
* `realization_controls` - control values for geo-realizations, also includes `simulation_id`
* `realization_objectives` - objective values per geo-realization
* `realization_constraints` - constraint values per geo-realization
* `perturbation_objectives` - objective and control values per perturbation
* `perturbation_constraints` - constraint and control values per perturbation (Note/discussion point: control values could be pulled into separate table to avoid redundancy)
* `batch_objective_gradient` - Partial derivatives of objectives, given different controls. This dataset has one column per objective, and one row per control value, and the intersecting cells represent the partial derivative of the objective wrt that control value.
* `batch_constraint_gradient` - Partial derivatives of constraints, given different controls. This dataset has one column per constraint, and one row per control value, and the intersecting cells represent the partial derivative of the constraint wrt that control value.

Example data from `math_func/config_advanced.yml` (json format)
Screenshot 2025-01-10 at 14 53 04

Exhaustive list of data stored **PER OPTIMIZATION** 
========================================
* `controls.json` - control values for this batch
* `realization_weights.json` - realization weights
* `nonlinear_constraints` - conditions for constraints to satisfy (on average over the batch)
* `objective_functions` - objective function names, weights, and normalization

Example data from `math_func/config_advanced.yml`
Screenshot 2025-01-10 at 15 00 29

Potential simplifications
==================
The `everest_data_api` is currently used for plotting, but could be used (probably expanded a bit) to avoid doing direct (polars) dataframe manipulations elsewhere in the code, but currently they are done directly in the code.

### CODE REVIEW

File: src/everest/api/everest_data_api.py [hunk 1]
```diff
@@ -1,46 +1,66 @@
...
+from everest.everest_storage import EverestStorage
 
 
 class EverestDataAPI:
     def __init__(self, config: EverestConfig, filter_out_gradient=True):
         self._config = config
         output_folder = config.optimization_output_dir
-        self._snapshot = SebaSnapshot(output_folder).get_snapshot(filter_out_gradient)
+        self._ever_storage = EverestStorage(Path(output_folder))
+        self._ever_storage.read_from_output_dir()
```
Comment (DanSava) [inline src/everest/api/everest_data_api.py:17]: This will load all the output data into memory. As far as I know, the everest outputs should not be too big even for large cases but I think it is something we need to keep an eye on.

File: src/everest/detached/__init__.py [hunk 2]
```diff
@@ -124,32 +124,63 @@ def wait_for_server(output_dir: str, timeout: int) -> None:
...
-        "control_history": snapshot.optimization_controls,
-        "objectives_history": snapshot.expected_objectives,
-        "accepted_control_indices": snapshot.increased_merit_indices,
+        "objective_history": expected_total_objective,
+        "control_history": polars.concat(
+            [
+                b.realization_controls.select(control_names)
+                for b in storage.data.batches
+                if b.realization_controls is not None
+            ]
```
Comment (DanSava) [inline src/everest/detached/__init__.py:180]: I guess there is a reason you decided not to move this type of logic inside `EverestStorageDataFrames`

**Review by DanSava — COMMENTED** [review]
This looks good :+1: :medal_sports:

File: src/everest/api/everest_data_api.py (hunk 1 above)
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py:17]: Good point. Yeah I think it is worth keeping an eye on, and I think it might make sense to do refactorings like making the results-fetching lazy-load later on, should not require a big change. It will also be more clear what to do when we know better how the data will be used/loaded.

File: src/everest/detached/__init__.py (hunk 2 above)
Comment (yngve-sk) [reply src/everest/detached/__init__.py:180]: These dataframe manipulations are candidates for some kind of refactoring for sure, but then we should put all the dataframe manipulation someplace central, I also think this can be a future issue/work that can be split up and done outside the scope of this PR.

File: src/everest/bin/config_branch_script.py [hunk 3]
```diff
@@ -46,10 +45,19 @@ def _build_args_parser():
...
-        if opt_data.batch_id == batch:
-            return opt_data.controls
+    storage = EverestStorage(Path(optimization_dir))
+    storage.read_from_output_dir()
+
+    control_names = storage.data.controls["control_name"]
+    batch_data = next((b for b in storage.data.batches if b.batch_id == batch), None)
+
+    if batch_data:
+        # All geo-realizations should have the same unperturbed control values per batch
```
Comment (StephanDeHoop) [inline src/everest/bin/config_branch_script.py:55]: Only true if batch contains a forward model run for unperturbed controls (if batch is only pertubations they will all be different for different `<GEO_ID>`s? Not sure how that affects `config_branch_entry()` where this is called.

File: tests/everest/entry_points/test_config_branch_entry.py [hunk 4]
```diff
@@ -28,19 +27,21 @@ def test_config_branch_entry(cached_example):
...
 
     new_controls_initial_guesses = {
         var[CK.INITIAL_GUESS] for var in new_controls[0][CK.VARIABLES]
     }
-    opt_control_val_for_batch_id = {v for k, v in opt_controls.items()}
 
-    assert new_controls_initial_guesses == opt_control_val_for_batch_id
+    control_names = storage.data.controls["control_name"]
+    batch_1_info = next(b for b in storage.data.batches if b.batch_id == 1)
+    realization_control_vals_mean = batch_1_info.realization_controls.select(
```
Comment (StephanDeHoop) [inline tests/everest/entry_points/test_config_branch_entry.py:39]: Why do we call it mean? It seems either unperturbed values or first perturbation or am I missing something?

File: tests/everest/entry_points/test_config_branch_entry.py [hunk 5]
```diff
@@ -80,5 +72,15 @@ def test_config_branch_preserves_config_section_order(cached_example):
 
     assert len(diff_lines) == 4
     assert "-initial_guess:0.25" in diff_lines
-    for control_val in opt_control_val_for_batch_id:
+
+    storage = EverestStorage(Path(path) / "everest_output" / "optimization_output")
+    storage.read_from_output_dir()
+    control_names = storage.data.controls["control_name"]
+    batch_1_info = next(b for b in storage.data.batches if b.batch_id == 1)
+    realization_control_vals_mean = batch_1_info.realization_controls.select(
```
Comment (StephanDeHoop) [inline tests/everest/entry_points/test_config_branch_entry.py:80]: Same as above

File: tests/everest/functional/test_main_everest_entry.py [hunk 6]
```diff
@@ -56,14 +57,15 @@ def test_everest_entry_run(cached_example):
 
     assert status["status"] == ServerStatus.completed
 
-    snapshot = SebaSnapshot(config.optimization_output_dir).get_snapshot()
+    storage = EverestStorage(Path(config.optimization_output_dir))
+    storage.read_from_output_dir()
+    optimal = storage.get_optimal_result()
```
Comment (StephanDeHoop) [inline tests/everest/functional/test_main_everest_entry.py:62]: Really like this (implicit?) improvement regarding the name compared to `snapshot.optimization_data[-1]` :) !

File: tests/everest/snapshots/test_api_snapshots/test_api_snapshots/config_multiobj.yml/snapshot.json [hunk 7]
```diff
@@ -69,27 +69,27 @@
     },
     {
       "batch": 0,
-      "control": "point_y",
-      "function": "distance_p",
-      "value": 0.98866227
+      "control": "point_x",
```
Comment (StephanDeHoop) [inline tests/everest/snapshots/test_api_snapshots/test_api_snapshots/config_multiobj.yml/snapshot.json:72]: Is this a weird git diff, or why did this control_name change in the test?

**Review by StephanDeHoop — COMMENTED** [review]
Looks great! Most of my comments are just asking to clarify things I don't understand :). Amazing work, well done :) !!!

File: src/everest/everest_storage.py [hunk 8]
```diff
@@ -0,0 +1,908 @@
...
+    realization_controls: polars.DataFrame
+    batch_objectives: polars.DataFrame | None
+    realization_objectives: polars.DataFrame | None
+    batch_constraints: polars.DataFrame | None
+    realization_constraints: polars.DataFrame | None
+    batch_objective_gradient: polars.DataFrame | None
+    perturbation_objectives: polars.DataFrame | None
+    batch_constraint_gradient: polars.DataFrame | None
+    perturbation_constraints: polars.DataFrame | None
+    is_improvement: bool | None = False
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:46]: Hmm, the `.existing_dataframes()` prop is used for looping over dataframes that exist for a batch, it may be function results and/or gradient results, so it filters out non-existent dataframes to avoid errors when trying to write them. As for other properties to the batch that are not dataframe, those will go into the `batch.json` file

File: src/everest/everest_storage.py [hunk 9]
```diff
@@ -0,0 +1,908 @@
...
+@dataclass
+class EverestStorageDataFrames:
+    batches: list[BatchDataFrames] = field(default_factory=list)
+    controls: polars.DataFrame | None = None
+    objective_functions: polars.DataFrame | None = None
+    nonlinear_constraints: polars.DataFrame | None = None
+    realization_weights: polars.DataFrame | None = None
+
+    @property
+    def simulation_to_geo_realization_map(self) -> dict[int, int]:
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:76]: I think this is a logic error now actually, since the mapping varies per `batch`, nice catch, will correct the logic

[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py]: Logic fixed, although this thing is not covered by tests. Suggest to defer adding a test for it until when we add those mappings explicitly to ERT storage, as it will be an easier setup than whatever is there now. There are probably some edge cases here I'm not quite familiar with.

File: src/everest/everest_storage.py [hunk 10]
```diff
@@ -0,0 +1,908 @@
...
+        to the user.
+        """
+        scaled_cols = [c for c in df.columns if c.lower().startswith("scaled")]
+        if len(scaled_cols) > 0:
+            raise ValueError("Scaled columns should not be stored into Everest storage")
+
+        renames = {
+            "objective": "objective_name",
+            "weighted_objective": "total_objective_value",
+            "variable": "control_name",
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:291]: Hmm, I think this rename function should clarify it somewhat. I added a comment clarifying that this maps ropt columns to columns we present to the user.

File: src/everest/everest_storage.py [hunk 11]
```diff
@@ -0,0 +1,908 @@
...
+    def _store_function_results(self, results: FunctionResults) -> _EvaluationResults:
+        # We could select only objective values,
+        # but we select all to also get the constraint values (if they exist)
+        realization_objectives = polars.from_pandas(
+            results.to_dataframe(
+                "evaluations",
+                select=["objectives", "evaluation_ids"],
+            ).reset_index(),
+        ).select(
+            "batch_id",
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:448]: Hmm, it would be possible to do an enum, but the column keys are still specified in the `_rename_ropt_df_columns` function, all other ROPT columns that are dropped, we do not need to know. We could maybe translate this into an enum but I'm not sure it would be a good practice / much gain, other than it being easier to get autocomplete while choosing a column. But I think if you are in the first place doing a dataframe manipulation, you will at the very least look at a sample dataframe before being able to make sense of what to do, so implicitly you'll know the columns in that way. If this becomes an issue we could do a future PR/issue towards it.

File: src/everest/everest_storage.py [hunk 12]
```diff
@@ -0,0 +1,908 @@
...
+            }
+        )
+
+        return df
+
+    def write_to_output_dir(self) -> None:
+        exp = _OptimizerOnlyExperiment(self._output_dir)
+
+        # csv writing mostly for dev/debugging/quick inspection
+        self.data.write_to_experiment(exp, write_csv=True)
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:345]: This was initially for dev/debug yes, and we did not discuss/land on a final format, I think this should be a discussion point at some point before merging it. The data is not big but I don't think we need to 3x the data with 3 different formats.

[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:345]: Added a commit making it optional for now, invocation now looks like this, but the whole json/csv should maybe be removed, to be discussed.

```
        self.data.write_to_experiment(
            exp,
            write_json=bool(os.getenv("_EVEREST_STORAGE_WRITE_JSON")),
            write_csv=bool(os.getenv("_EVEREST_STORAGE_WRITE_CSV")),
        )
```

[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py]: Edit: Added another commit on top removing the functionality, I think we should probably default to not having it as a feature.

File: src/everest/everest_storage.py [hunk 13]
```diff
@@ -0,0 +1,908 @@
...
+    def get_optimal_result(self) -> OptimalResult | None:
+        # Only used in tests, but re-created to ensure
+        # same behavior as w/ old SEBA setup
+        has_merit = any(
+            "merit_value" in b.batch_objectives.columns
+            for b in self.data.batches
+            if b.batch_objectives is not None
+        )
+
+        def find_best_batch(
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:792]: I'm not sure that we do, though it is part of the whole "reporting results in a nice and non-misleading way to the user" part of this.

File: src/everest/everest_storage.py [hunk 14]
```diff
@@ -0,0 +1,908 @@
...
+                    polars.col("total_objective_value").sample(n=1)
+                ).item(),
+            )
+        else:
+            # Maximize objective
+            batch, controls_dict = find_best_batch(
+                filter_by=lambda b: b.batch_objectives is not None
+                and not b.batch_objectives.is_empty(),
+                sort_by=lambda b: -b.batch_objectives.select(
+                    polars.col("total_objective_value").sample(n=1)
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:840]: AFAIK, the `batch_objectives` will contain only one row for the one batch, one column denoting the `total_objective_value`, and one col per response, so this should be OK. Agree @verveerpj ?

Ref:
```
self.data.batches[1].batch_objectives
Out[4]: 
shape: (1, 4)
┌──────────┬───────────────────────┬────────────┬────────────┐
│ batch_id ┆ total_objective_value ┆ distance_p ┆ distance_q │
│ ---      ┆ ---                   ┆ ---        ┆ ---        │
│ u32      ┆ f64                   ┆ f64        ┆ f64        │
╞══════════╪═══════════════════════╪════════════╪════════════╡
│ 1        ┆ -2.333526             ┆ -0.765646  ┆ -4.70364   │
└──────────┴───────────────────────┴────────────┴────────────┘
self.data.batches[2].batch_objectives
Out[5]: 
shape: (1, 4)
┌──────────┬───────────────────────┬────────────┬────────────┐
│ batch_id ┆ total_objective_value ┆ distance_p ┆ distance_q │
│ ---      ┆ ---                   ┆ ---        ┆ ---        │
│ u32      ┆ f64                   ┆ f64        ┆ f64        │
╞══════════╪═══════════════════════╪════════════╪════════════╡
│ 2        ┆ -2.000048             ┆ -0.507785  ┆ -4.47679   │
└──────────┴───────────────────────┴────────────┴────────────┘
```

File: src/everest/everest_storage.py [hunk 15]
```diff
@@ -0,0 +1,908 @@
...
+            )
+
+            if batch is None:
+                return None
+
+            return OptimalResult(
+                batch=batch.batch_id,
+                controls=controls_dict,
+                total_objective=batch.batch_objectives.select(
+                    polars.col("total_objective_value")
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:851]: Same answer as above

File: src/everest/api/everest_data_api.py [hunk 16]
```diff
@@ -1,46 +1,66 @@
...
+        initial_values = self._ever_storage.data.controls
+        control_spec = initial_values.filter(
+            pl.col("control_name") == control
+        ).to_dicts()[0]
+        return {
+            "min": control_spec.get("lower_bounds"),
+            "max": control_spec.get("upper_bounds"),
+        }
 
     def output_constraint(self, constraint):
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py:65]: Only legacy, only referred to in the snapshot test, might (very likely) be removed in future, but for now I think we just port over behavior, and then do the removals incrementally afterwards.

File: src/everest/api/everest_data_api.py [hunk 17]
```diff
@@ -50,106 +70,128 @@ def output_constraint(self, constraint):
                  "right_hand_side" is a constant real number that indicates
                  the constraint bound/target.
         """
-        constraints = [
-            con
-            for con in self._snapshot.metadata.constraints.values()
-            if con.name == constraint
-        ]
+
+        constraint_dict = self._ever_storage.data.nonlinear_constraints.to_dicts()[0]
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py:74]: Oh, this is a logic error, now corrected. It should select the constraint name provided in the arg here.

File: src/everest/api/everest_data_api.py [hunk 18]
```diff
@@ -50,106 +70,128 @@ def output_constraint(self, constraint):
...
         return {
-            "type": constraints[0].constraint_type,
-            "right_hand_side": constraints[0].rhs_value,
+            "type": ConstraintType(constraint_dict["constraint_type"]).name.lower(),
+            "right_hand_side": constraint_dict["constraint_rhs_value"],
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py:79]: Probably no good reason for it, legacy. If we want the forward model to produce `N < 10` RHS would be 10 and type would be some formulation of "less than". Renaming to something else can be discussed

File: src/everest/api/everest_data_api.py [hunk 19]
```diff
@@ -50,106 +70,128 @@ def output_constraint(self, constraint):
...
         }
 
     @property
     def realizations(self):
-        return list(
-            OrderedDict.fromkeys(
-                int(sim.realization) for sim in self._snapshot.simulation_data
-            )
+        return sorted(
+            self._ever_storage.data.batches[0]
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py:85]: Yeah this is weird, it might be unperturbed evals only also. Logic should be revised, but maybe in the future when we better see the usages of these endpoints. The first batch will always be of unperturbed georealizations, never gradient, I think. Agree @verveerpj ?

[the remark this answers is not in the document]
Comment (verveerpj) [reply src/everest/api/everest_data_api.py:85]: Indeed, it is possible that an optimization algorithm starts with a gradient. But currently, that would always trigger a unperturbed function evaluation, so its safe.

[the remark this answers is not in the document]
Comment (verveerpj) [reply src/everest/api/everest_data_api.py:85]: Indeed, it is possible that an optimization algorithm starts with a gradient. But currently, that would always trigger a unperturbed function evaluation, so its safe.

[the remark this answers is not in the document]
Comment (verveerpj) [reply src/everest/api/everest_data_api.py:85]: Indeed, it is possible that an optimization algorithm starts with a gradient. But currently, that would always trigger a unperturbed function evaluation, so its safe.

File: src/everest/everest_storage.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (verveerpj) [reply src/everest/everest_storage.py:840]: Each batch should have only one total objective and one value for each individual objective (or constraint), since these are aggregated over the realizations. So that translates to a single row per batch, I assume.

File: src/everest/api/everest_data_api.py [hunk 20]
```diff
@@ -50,106 +70,128 @@ def output_constraint(self, constraint):
...
-                "realization": sim.realization,
-                "simulation": sim.simulation,
-                "value": sim.objectives[objective.name],
-                "weight": objective.weight,
-                "norm": objective.normalization,
-            }
-            for sim in self._snapshot.simulation_data
-            for objective in self._snapshot.metadata.objectives.values()
-            if objective.name in sim.objectives
+            b for b in self._ever_storage.data.batches if b.batch_objectives is not None
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py]: Fixed, added logic for this endpoint. It only changes the ordering of results, but not results in themselves. Depends on https://github.com/equinor/ert/actions/runs/12828815639?pr=9791 being in first (PR is currently rebased on top of this though)

File: src/everest/everest_storage.py [hunk 21]
```diff
@@ -0,0 +1,908 @@
...
+    controls: list[Any]
+    total_objective: float
+
+
+def try_read_df(path: Path) -> polars.DataFrame | None:
+    return polars.read_parquet(path) if path.exists() else None
+
+
+@dataclass
+class BatchDataFrames:
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/everest_storage.py:35]: Hmm, it was intended to be more explicit than just `BatchData`. I renamed them to `BatchStorageData` and `OptimizationStorageData` to reflect a bit better what kind of data is stored, can be discussed/reviewed still. Main thing is the batch data is per batch, and optimization is all the batch data + some optimization-wide data (which may or may not be dropped in the future, then we might revise this once more)

File: src/everest/api/everest_data_api.py [hunk 22]
```diff
@@ -180,16 +222,13 @@ def summary_values(self, batches=None, keys=None):
...
-                # The realization ID as defined by Everest must be
-                # retrieved via the seba snapshot.
-                realization_map = {
-                    sim.simulation: sim.realization
-                    for sim in self._snapshot.simulation_data
-                    if sim.batch == batch_id
-                }
+
+                realization_map = (
+                    self._ever_storage.data.simulation_to_geo_realization_map
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/everest/api/everest_data_api.py]: this is `geo_realization`, and is now also made batch dependent, I think addressing your previous comment also takes this

File: src/everest/bin/config_branch_script.py (hunk 3 above)
Comment (yngve-sk) [reply src/everest/bin/config_branch_script.py:56]: Hmm, I think this is OK in that case, because it is used for `config_branch` functionality, where users supply a batch to use as a "prior" in the next config, basically copying its control values as initial values in the new config. If they supply a perturbation-only batch there will be some (probably not the best) error, but I think that is separate from this PR.

File: tests/everest/entry_points/test_config_branch_entry.py (hunk 4 above)
Comment (yngve-sk) [reply tests/everest/entry_points/test_config_branch_entry.py:38]: Oh, I think it is because it was maybe running on advanced before, which had 2 geo-realizations, where it did the mean, but on minimal when there is only one, the mean is no longer necessary.

File: tests/everest/entry_points/test_config_branch_entry.py (hunk 5 above)
Comment (yngve-sk) [reply tests/everest/entry_points/test_config_branch_entry.py:79]: Same as above, renaming the variable to not have `mean` in it

File: tests/everest/snapshots/test_api_snapshots/test_api_snapshots/config_multiobj.yml/snapshot.json (hunk 7 above)
Comment (yngve-sk) [reply tests/everest/snapshots/test_api_snapshots/test_api_snapshots/config_multiobj.yml/snapshot.json:72]: Worth a double check, but I think it just means the order of items changed, and the diff goes line by line rather than "by order"

File: src/everest/everest_storage.py [hunk 23]
```diff
@@ -0,0 +1,897 @@
...
+        self.data.write_to_experiment(exp)
+
+    def read_from_output_dir(self) -> None:
+        exp = _OptimizerOnlyExperiment(self._output_dir)
+        self.data.read_from_experiment(exp)
+
+    def observe_optimizer(
+        self,
+        optimizer: BasicOptimizer,
+        merit_file: Path,
```
Comment (verveerpj) [inline src/everest/everest_storage.py:341]: I would not pass the merit file path as an argument. As it is, this is an implementation detail to get merit handling to work with one specific dakota algorithm. I would hard-code somewhere in the merit-handling code, and get rid of the `self._merit` attribute. We need to remove this merit handling approach in the near future anywhen when we do better handling of constraint violations.

File: src/everest/everest_storage.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (StephanDeHoop) [reply src/everest/everest_storage.py]: Sounds good :)!

File: src/everest/everest_storage.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (StephanDeHoop) [reply src/everest/everest_storage.py]: Yes I agree.

File: src/everest/everest_storage.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (StephanDeHoop) [reply src/everest/everest_storage.py:788]: Ah ok, I will discuss with Pieter, it's definitely not part of this story/pr, but I'm interested to know.

File: src/everest/api/everest_data_api.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (StephanDeHoop) [reply src/everest/api/everest_data_api.py:80]: Ah, ok I understand, no leaving it as is is fine for me.

File: src/everest/everest_storage.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (StephanDeHoop) [reply src/everest/everest_storage.py]: Makes sense, looks good to me now :)!

File: tests/everest/entry_points/test_config_branch_entry.py (hunk 4 above)
Comment (StephanDeHoop) [reply tests/everest/entry_points/test_config_branch_entry.py]: Oef, then maybe I should have changed this (I probably changed the config file that was used when I was simplifying the tests haha), thanks :)!

File: src/everest/everest_storage.py [hunk 24]
```diff
@@ -0,0 +1,893 @@
...
+                values=[
+                    "constraint_value",
+                    "constraint_violation",
+                ],
+                separator=";",
+            ).rename(
+                {
+                    **{f"constraint_value;{name}": name for name in constraint_names},
+                    **{
+                        f"constraint_violation;{name}": f"{name}.violation"
```
Comment (verveerpj) [inline src/everest/everest_storage.py:479]: I think it would be better not to access the `violation` fields at the moment. They are used by `ropt` to report constraint violations, but that is not used by Everest. I think I will make some changes to `ropt` in the way these are reported, so lets not use this for now.

File: src/everest/everest_storage.py [hunk 25]
```diff
@@ -0,0 +1,880 @@
...
+                "constraints",
+            )
+
+            realization_constraints = self._rename_ropt_df_columns(
+                realization_constraints
+            )
+
+            batch_constraints = polars.from_pandas(
+                results.to_dataframe("nonlinear_constraints").reset_index()
+            ).select("batch_id", "nonlinear_constraint", "values")
```
Comment (verveerpj) [inline src/everest/everest_storage.py:456]: Don't use the `nonlinear_constraints` field. It is used to store the difference between the constraint and its rhs value, not the total constraint itself. Also, this field is likely to disappear, since current contraint handling is to be revised in `ropt`.

Instead, get the values from the `functions` fields, see comment below.

File: src/everest/everest_storage.py [hunk 26]
```diff
@@ -0,0 +1,880 @@
...
+        else:
+            batch_constraints = None
+            realization_constraints = None
+
+        batch_objectives = polars.from_pandas(
+            results.to_dataframe(
+                "functions",
+                select=["objectives", "weighted_objective"],
+            ).reset_index()
+        ).select("batch_id", "objective", "objectives", "weighted_objective")
```
Comment (verveerpj) [inline src/everest/everest_storage.py:484]: You can grab `constraints` from here.

File: src/everest/everest_storage.py [hunk 27]
```diff
@@ -0,0 +1,880 @@
...
+            ).reset_index(),
+        ).select(
+            "batch_id",
+            "realization",
+            "objective",
+            "objectives",
+            "evaluation_ids",
+        )
+
+        if results.nonlinear_constraints is not None:
```
Comment (verveerpj) [inline src/everest/everest_storage.py:436]: This is a bug, it should be:
`if results.functions is not None and results.functions.constraints is not None:`

**Review by verveerpj — APPROVED** [review]
One bug to fix, and it should be good.

File: src/everest/everest_storage.py (hunk 27 above)
Comment (yngve-sk) [reply src/everest/everest_storage.py:436]: fixed

File: src/everest/everest_storage.py [hunk 28]
```diff
@@ -0,0 +1,895 @@
...
+        if os.path.exists(self._output_dir / "seba.db") or os.path.exists(
+            self._output_dir / "seba.db.backup"
+        ):
+            logging.getLogger(EVEREST).error(
+                f"Tried opening old seba storage."
+                f"Traceback: {'\n'.join(traceback.format_stack())}"
+            )
+            raise SystemExit(
+                f"Trying to open old storage @ {self._output_dir}/seba.db."
+                f"This storage can only be opened with an ERT version <= 12.1.2."
```
Comment (verveerpj) [inline src/everest/everest_storage.py:346]: Maybe say `Everest/ERT`?
